BibtexQuery/Format.lean

```diff
@@ -172,133 +172,191 @@ partial def getLastCharOfContent (x : Content) : Option Char :=
 
-def sentence1 (arr : Array Content) : Option (Array Content) :=
+def sentence1 (arr : Array Content) : Array Content :=
   if arr.isEmpty then
-    .none
-  else if GeneralCategory.isPunctuation (getLastCharOfArrayContent arr |>.getD default) then
-    .some arr
+    #[]
+  else if canEndSentence (getLastCharOfArrayContent arr |>.getD ' ') then
+    arr
   else
-    .some (arr ++ #[.Character "."])
+    arr ++ #[.Character "."]
 
-def sentence (arr : Array (Option (Array Content))) : Option (Array Content) :=
+def sentence (arr : Array (Option (Array Content))) : Array Content :=
   sentence1 <| arrayConcat arr #[.Character ", "]
 
-def sentence' (s : String) : Array Content :=
-  if s.isEmpty then
-    #[]
-  else if GeneralCategory.isPunctuation s.back then
-    #[.Character s]
-  else
-    #[.Character (s ++ ".")]
+def sentence' (content : Array TexContent) : Array Content :=
+  sentence1 <| TexContent.toHtmlArray content
 
 def toplevel (arr : Array (Option (Array Content))) : Array Content :=
   arrayConcat arr #[.Character "\n"]
 
+def mkUrl (url : String) : Array Content :=
+  if url.isEmpty then #[] else #[.Character "URL: "] ++ mkHref url #[.normal url]
+
+def mkWebRef (urlPrefix namePrefix url : String) : Array Content :=
+    let s :=
+      if url.toLower.startsWith urlPrefix.toLower then
+        url.drop urlPrefix.length
+      else if url.toLower.startsWith namePrefix.toLower then
+        url.drop namePrefix.length
+      else
+        url
+    if s.toLower.startsWith "http" then
+      -- the url does not starts with `urlPrefix` or `namePrefix`,
+      -- but still starts with "http"
+      mkUrl url
+    else if s.isEmpty then
+      #[]
+    else
+      mkHref (urlPrefix ++ s) #[.normal (namePrefix ++ s)]
+
+/-!
+
+### Partial templates
+
+-/
+
 def formatVolumeAndPages : Option (Array Content) := do
   let pages ← e.tags["pages"]?
   match e.tags["volume"]? with
   | .some volume =>
     arrayConcat #[
-      mkStr' volume,
+      mkStr volume,
       (mkStr · "(" ")") <$> e.tags["number"]?,
-      mkStr' pages ":"
+      mkStr pages ":"
     ]
-  | .none => mkStr' pages "pages "
-
-def formatWebRefs : Option (Array Content) :=
-  let formatUrl : Option (Array Content) := do
-    let s ← e.tags["url"]?
-    pure <| if s.isEmpty then #[] else #[.Character "URL: ", mkHref s s]
-  let formatUrl' (tagName urlPrefix namePrefix : String) : Option (Array Content) := do
-    let s ← e.tags[tagName]?
-    pure <| mkHref' (urlPrefix ++ s) (namePrefix ++ s)
+  | .none => mkStr pages "pages "
+
+def formatEprint : Option (Array Content) := do
+  let eprint ← TexContent.toPlaintextArray <$> e.tags["eprint"]?
+  let eprinttype :=
+    TexContent.toPlaintextArray <$> (e.tags["eprinttype"]? <|> e.tags["archiveprefix"]?)
+      |>.getD "arXiv"
+  let eprintlist : Array (Array String × String × String) := #[
+    (#["arxiv"], "https://arxiv.org/abs/", "arXiv:"),
+    (#["pmcid", "pmc"], "https://www.ncbi.nlm.nih.gov/pmc/articles/", "PMCID:")
```

Comment:
```suggestion
    (#["arxiv"], "https://arxiv.org/abs/", "arXiv:"),
    (#["iacr"], "https://eprint.iacr.org/", "IACR:"),
    (#["pmcid", "pmc"], "https://www.ncbi.nlm.nih.gov/pmc/articles/", "PMCID:")
```
Can you add this one as well? It's the IACR (International Association for Cryptologic Research), which runs its own eprint server for cryptography papers.

Reply:
Sure.